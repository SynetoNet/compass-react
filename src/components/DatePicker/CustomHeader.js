import React from "react";

import { getMonth, getYear } from "date-fns";
import range from "lodash/range";
import uniq from "lodash/uniq";

const months = [
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December"
];

const CustomHeader = ({
  date,
  changeYear,
  changeMonth,
  decreaseMonth,
  increaseMonth,
  prevMonthButtonDisabled,
  nextMonthButtonDisabled,
  ...props
}) => {
  const years = getYearsRange(props);

  return (
    <div
      style={{
        margin: 20,
        display: "flex",
        justifyContent: "center"
      }}
    >
      <button
        className="react-datepicker__navigation react-datepicker__navigation--previous"
        onClick={decreaseMonth}
        disabled={prevMonthButtonDisabled}
      >
        <span className="react-datepicker__navigation-icon react-datepicker__navigation-icon--previous">Previous Month</span>
      </button>
      <select
        value={getYear(date)}
        onChange={({ target: { value } }) => changeYear(value)}
        className="datepicker-dropdown datepicker-year-dropdown"
      >
        {years.map(option => (
          <option key={option} value={option}>
            {option}
          </option>
        ))}
      </select>

      <select
        value={getMonth(date)}
        onChange={({ target: { value } }) => changeMonth(value)}
        className="datepicker-dropdown datepicker-month-dropdown"
      >
        {months.map((month, index) => (
          <option key={index} value={index}>
            {month}
          </option>
        ))}
      </select>

      <button
        className="react-datepicker__navigation react-datepicker__navigation--next"
        onClick={increaseMonth}
        disabled={nextMonthButtonDisabled}
      >
        <span className="react-datepicker__navigation-icon react-datepicker__navigation-icon--next">Next Month</span>
      </button>
    </div>
  );
};

export default CustomHeader;

function getYearsRange({ minDate, maxDate, includeDates }) {
  if (includeDates) {
    return uniq(includeDates.map(getYear)).sort();
  }

  let startYear = 2000;
  let endYear = getYear(new Date()) + 10;

  if (minDate) {
    startYear = Math.max(startYear, getYear(minDate));
  }
  if (maxDate) {
    endYear = Math.min(endYear, getYear(maxDate));
  }

  return range(startYear, endYear, 1);
}
