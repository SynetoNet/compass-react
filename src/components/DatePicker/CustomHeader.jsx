import React from "react";

import getMonth from "date-fns/getMonth";
import getYear from "date-fns/getYear";
import range from "lodash/range";

const years = range(1960, getYear(new Date()) + 20, 1);
const months = Object.entries({
  0: "January",
  1: "February",
  2: "March",
  3: "April",
  4: "May",
  5: "June",
  6: "July",
  7: "August",
  8: "September",
  9: "October",
  10: "November",
  11: "December"
});

const CustomHeader = ({
  date,
  changeYear,
  changeMonth,
  decreaseMonth,
  increaseMonth,
  prevMonthButtonDisabled,
  nextMonthButtonDisabled
}) => {
  return (
    <div
      style={{
        margin: 20,
        display: "flex",
        justifyContent: "center"
      }}
    >
      <button
        className="react-datepicker__navigation react-datepicker__navigation--previous"
        onClick={decreaseMonth}
        disabled={prevMonthButtonDisabled}
      />
      <select
        value={getYear(date)}
        onChange={({ target: { value } }) => changeYear(value)}
        className="datepicker-dropdown datepicker-year-dropdown"
      >
        {years.map(option => (
          <option key={option} value={option}>
            {option}
          </option>
        ))}
      </select>

      <select
        value={getMonth(date)}
        onChange={({ target: { value } }) => changeMonth(value)}
        className="datepicker-dropdown datepicker-month-dropdown"
      >
        {months.map(month => (
          <option key={month[0]} value={month[0]}>
            {month[1]}
          </option>
        ))}
      </select>

      <button
        className="react-datepicker__navigation react-datepicker__navigation--next"
        onClick={increaseMonth}
        disabled={nextMonthButtonDisabled}
      />
    </div>
  );
};

export default CustomHeader;
