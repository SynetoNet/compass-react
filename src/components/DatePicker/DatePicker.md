#### **Default usage**

This is a custom version of [react-datepicker](https://reactdatepicker.com).

By default, **`DatePicker`** should be treated as a controlled component, by binding the **`selected`** value and the **`onChange`** handler.

DatePicker will render a **`<Form.Control />`** element by default.

```jsx
import { useState } from "react";

const DefaultExample = () => {
  const [date, setDate] = useState(new Date());

  return <DatePicker selected={date} onChange={val => setDate(val)} />;
};

<DefaultExample />;
```

#### **Inline calendar**

```jsx
import { useState } from "react";

const InlineExample = () => {
  const [date, setDate] = useState(new Date());

  return <DatePicker selected={date} onChange={val => setDate(val)} inline />;
};

<InlineExample />;
```

#### **Date format**

We use [date-fns](https://date-fns.org/) to format date, see [documentation](https://date-fns.org/v2.0.0-alpha.23/docs/format) for details.

---

> **IMPORTANT NOTE**: This doesn't work with **`redux-form`**, because the update is handled outside this component.

> You should manually format the **initial value** and the **updated value**.

---

```jsx
import { useState } from "react";

const DefaultExample = () => {
  const [date, setDate] = useState(new Date());

  return (
    <DatePicker
      selected={date}
      onChange={val => setDate(val)}
      dateFormat="dd-MMM-yyyy h:mm aa"
    />
  );
};

<DefaultExample />;
```

#### **Custom input**

If you use another library for form handling, like **Formik** or **redux-forms**, you can pass a custom element.

```jsx
import { useState } from "react";
import { Form } from "react-bootstrap";

const CustomInputExample = () => {
  const [date, setDate] = useState("");

  return (
    <DatePicker
      selected={date}
      onChange={val => setDate(val)}
      customInput={
        <Form.Control
          isValid={!!date}
          isInvalid={!date}
          type="text"
          placeholder="You must select a date"
        />
      }
      dateFormat="dd MMM, yyyy h:mm aa"
    />
  );
};

<CustomInputExample />;
```

#### **Scrollable Modal example**

Please use the **`enforceFocus={ false }`** prop on the Modal component, otherwise the month and year dropdowns may have unwanted behavior

```jsx
import { useState } from "react";
import { Form } from "react-bootstrap";
import Button from "../Button/Button";
import Modal from "../Modal/Modal";

const ExampleModal = () => {
  const [showModal, toggleModal] = useState(false);
  const [date, setDate] = useState(new Date());

  return (
    <>
      <Button variant="primary" onClick={() => toggleModal(true)}>
        Open modal
      </Button>

      <Modal
        show={showModal}
        onHide={() => toggleModal(false)}
        scrollable
        enforceFocus={ false }
        content={
          <div>
            <p>Scroll down</p>
            <div style={{ height: "100vh" }} />
            <DatePicker
              selected={date}
              onChange={val => setDate(val)}
              customInput={<Form.Control type="text" />}
            />
            <div style={{ height: "100vh" }} />
          </div>
        }
        title="Modal title"
      />
    </>
  );
};

<ExampleModal />;
```

#### **Highlight specific dates**

```jsx
import { useState } from "react";
import { tomorrow, yesterday } from "./example.utils.js";

const HighlightExample = () => {
  const [date, setDate] = useState(new Date());

  return (
    <DatePicker
      selected={date}
      onChange={val => setDate(val)}
      highlightDates={[tomorrow, yesterday]}
    />
  );
};

<HighlightExample />;
```

#### **Placeholder**

```jsx
import { useState } from "react";

const PlaceholderExample = () => {
  const [date, setDate] = useState("");

  return (
    <DatePicker
      selected={date}
      onChange={val => setDate(val)}
      placeholderText="Please select a date"
    />
  );
};

<PlaceholderExample />;
```

#### **Filter available dates**

**Minimum date:**

```jsx
import { useState } from "react";
import { yesterday } from "./example.utils.js";

const MinDateExample = () => {
  const [date, setDate] = useState(new Date());

  return (
    <DatePicker
      selected={date}
      onChange={val => setDate(val)}
      minDate={yesterday}
    />
  );
};

<MinDateExample />;
```

**Maximum date:**

```jsx
import { useState } from "react";
import { tomorrow } from "./example.utils.js";

const MaxDateExample = () => {
  const [date, setDate] = useState(new Date());

  return (
    <DatePicker
      selected={date}
      onChange={val => setDate(val)}
      maxDate={tomorrow}
    />
  );
};

<MaxDateExample />;
```

**Include specific dates:**

```jsx
import { useState } from "react";
import { tomorrow, today, yesterday } from "./example.utils.js";

const IncludeDatesExample = () => {
  const [date, setDate] = useState(new Date());

  return (
    <DatePicker
      selected={date}
      onChange={val => setDate(val)}
      includeDates={[tomorrow, today, yesterday]}
    />
  );
};

<IncludeDatesExample />;
```

**Exclude specific dates:**

```jsx
import { useState } from "react";
import { tomorrow, yesterday } from "./example.utils.js";

const ExcludeDatesExample = () => {
  const [date, setDate] = useState(new Date());

  return (
    <DatePicker
      selected={date}
      onChange={val => setDate(val)}
      excludeDates={[tomorrow, yesterday]}
    />
  );
};

<ExcludeDatesExample />;
```

**Filter out dates based on a rule:**

```jsx
import { useState } from "react";

function isWeekday(dateObject) {
  const weekDay = dateObject.getDay();
  return weekDay > 0 && weekDay < 6;
}

const FilterDatesExample = () => {
  const [date, setDate] = useState(new Date());

  return (
    <DatePicker
      selected={date}
      onChange={val => setDate(val)}
      filterDate={isWeekday}
      placeholderText="Select weekdays only "
    />
  );
};

<FilterDatesExample />;
```

#### **Time picker**

For a comprehensive list of options, please checkout [react-datepicker's docs](https://reactdatepicker.com).

**Date & Time picker**

```jsx
import { useState } from "react";

const TimeExample = () => {
  const [date, setDate] = useState(new Date());

  return (
    <DatePicker
      selected={date}
      showTimeSelect
      onChange={val => setDate(val)}
      dateFormat="dd-MMM-yyyy hh:mm"
    />
  );
};

<TimeExample />;
```

**Time picker only**

```jsx
import { useState } from "react";

const TimeExample = () => {
  const [date, setDate] = useState(new Date());

  return (
    <DatePicker
      selected={date}
      showTimeSelect
      showTimeSelectOnly
      onChange={val => setDate(val)}
      dateFormat="hh:mm"
    />
  );
};

<TimeExample />;
```

**Control time intervals**

```jsx
import { useState } from "react";

const EveryHour = () => {
  const [date, setDate] = useState("");

  return (
    <DatePicker
      selected={date}
      showTimeSelect
      showTimeSelectOnly
      timeIntervals={60}
      onChange={val => setDate(val)}
      placeholderText="Select time (1 hour increments)"
    />
  );
};

<EveryHour />;
```
