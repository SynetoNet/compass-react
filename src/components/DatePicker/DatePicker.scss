@import "../../base/colors";
@import "../../base/variables";

.popover-datepicker {
  max-width: none !important;
  background-color: transparent;
  border: none;

  .popover-body {
    padding: 0;
  }

  .arrow {
    display: none;
  }
}

.react-datepicker__input-container,
.react-datepicker-wrapper {
  display: block;
}

.react-datepicker-popper[data-placement^="bottom"] {
  margin-top: 0;
}

.react-datepicker-popper[data-placement^="top"] {
  top: 4px !important;
  margin-bottom: 0;
}

.react-datepicker {
  font-family: Roboto, sans-serif;
  font-weight: normal;
  font-size: $font-size-base;
  border-radius: 2px;
  border: 1px solid $neutral-12;
  box-shadow: 0 2px 10px 2px $neutral-12;

  // display: inline-flex;
}

.react-datepicker__month-container {
  padding: 0 8px 8px;

  // float: none;
}

.react-datepicker__time-container {
  // float: none;
  width: auto;

  // padding: 0px 8px 8px;

  .react-datepicker__time .react-datepicker__time-box {
    width: auto;

    // padding-top: 8px;
  }

  .react-datepicker__time-list-item:hover:not(.react-datepicker__time-list-item--selected) {
    background-color: $accent-light-85 !important;
  }

  .react-datepicker__time-list-item--selected {
    background-color: $primary !important;
  }

  .react-datepicker-time__header {
    padding: 0.5rem 0;
  }
}

.react-datepicker__header {
  background: $neutral-0;
  border-bottom: 1px solid $neutral-12;
  padding: 0;

  .react-datepicker__current-month {
    color: $neutral-90;
    margin-bottom: 11px;
  }
}

.react-datepicker__month {
  margin-top: 8px;
  color: $neutral-90;
}

.react-datepicker__day--outside-month {
  visibility: hidden;
}

.react-datepicker__day {
  padding: 3px;
  margin: 1px;
  width: 36px;
  line-height: 30px;
  position: relative;
  color: $neutral-90;

  &:hover {
    border-radius: 1px;
    background-color: $accent-light-85;
    color: $neutral-90;
  }

  &.react-datepicker__day--disabled {
    opacity: 0.5;

    &:hover {
      background-color: $neutral-0;
    }
  }
}

.react-datepicker__day-names {
  display: flex;
  justify-content: space-evenly;
  margin: 0;
}

.react-datepicker__day--today {
  background-color: $neutral-0;
  border-radius: 1px;
  color: $neutral-90;
  position: relative;

  &.react-datepicker__day--selected::after {
    position: absolute;
    content: "";
    border: 2px solid $neutral-0;
    border-radius: 2px;
    bottom: 6px;
    right: 45%;
  }
}

.react-datepicker__day--today::after {
  position: absolute;
  content: "";
  border: 2px solid $primary;
  border-radius: 2px;
  bottom: 6px;
  right: 45%;
}

.react-datepicker__day--highlighted {
  background: none;
}

.react-datepicker__day--highlighted::after {
  position: absolute;
  content: "";
  border: 1px solid $neutral-90;
  border-radius: 1px;
  bottom: 8px;
  right: 45%;
}

.react-datepicker__day--selected {
  background-color: $primary;
  border-radius: 1px;
  color: $neutral-0;

  &:hover {
    background-color: $primary;
    border-radius: 1px;
    color: $neutral-0;
  }

  &.react-datepicker__day--disabled {
    opacity: 0.5;
    background-color: $neutral-0;
    color: #2e324b;
    font-weight: normal;

    &:hover {
      background-color: $neutral-0;
    }
  }
}

.react-datepicker__navigation {
  outline: none;
  top: 0;
}

.react-datepicker__navigation--next {
  border-left-color: $neutral-60;
}

.react-datepicker__navigation--previous {
  border-right-color: $neutral-60;
}

.react-datepicker__triangle {
  display: none;
}

.datepicker-dropdown {
  background: none;
  outline: none;
  border: none;

  &.datepicker-year-dropdown {
    margin-right: 10px;
  }
}
