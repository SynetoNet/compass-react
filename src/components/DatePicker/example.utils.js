export const today = new Date();
export const tomorrow = new Date();
export const yesterday = new Date();

tomorrow.setDate(today.getDate() + 1);
yesterday.setDate(today.getDate() - 1);
