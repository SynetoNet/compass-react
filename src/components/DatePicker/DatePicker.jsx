import React, { useState, useRef } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import ReactDatePicker from "react-datepicker";
import { Overlay, Popover } from "react-bootstrap";

import "react-datepicker/dist/react-datepicker.css";
import "./DatePicker.scss";

const DatePicker = props => {
  const [showDatePicker, toggleDatePicker] = useState(false);
  const inputRef = useRef(null);
  const classes = classNames({});

  if (props.customInput) {
    const CustomInput = React.cloneElement(
      props.customInput,
      {
        onFocus: () => toggleDatePicker(true),
        ref: inputRef,
        onChange: () => {},
        value: props.selected
      },
      null
    );

    return (
      <>
        {CustomInput}
        <Overlay
          target={inputRef.current}
          show={showDatePicker}
          placement="auto"
        >
          <Popover className="popover-datepicker">
            <ReactDatePicker
              inline
              className={classes}
              onSelect={val => {
                props.onChange(val);
                toggleDatePicker(false);
              }}
              onClickOutside={() => toggleDatePicker(false)}
              {...props}
            />
          </Popover>
        </Overlay>
      </>
    );
  }

  return <ReactDatePicker className={classes} {...props} />;
};

DatePicker.propTypes = {
  /** You must pass a __`Date`__, not a `string` */
  selected: PropTypes.oneOfType([
    PropTypes.oneOf(["", undefined]),
    PropTypes.instanceOf(Date)
  ]),
  onChange: PropTypes.func.isRequired,
  /**
   * You can pass your custom input element
   * (otherwise it will render a default `<input />` tag)
   */
  customInput: PropTypes.element,
  placeholderText: PropTypes.string,
  /** You must pass an __`Array<Date>`__ */
  highlightDates: PropTypes.arrayOf(PropTypes.instanceOf(Date)),
  /** You must pass a __`Date`__ */
  minDate: PropTypes.instanceOf(Date),
  /** You must pass a __`Date`__ */
  maxDate: PropTypes.instanceOf(Date),
  /** You must pass an __`Array<Date>`__ */
  includeDates: PropTypes.arrayOf(PropTypes.instanceOf(Date)),
  /** You must pass an __`Array<Date>`__ */
  excludeDates: PropTypes.arrayOf(PropTypes.instanceOf(Date)),
  /**
   * Filter function, return __`true`__ to enable date, __`false`__ to disable
   *
   * @param {Object} currentDay
   */
  filterDate: PropTypes.func
};

DatePicker.defaultProps = {
  selected: undefined
};

export default DatePicker;
