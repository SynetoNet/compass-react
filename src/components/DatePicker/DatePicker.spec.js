import React from "react";
import DatePicker from "./DatePicker";
import renderer from "react-test-renderer";

test("renders default DatePicker", () => {
  const Component = () => {
    return <DatePicker selected={new Date()} onChange={() => {}} />;
  };

  const component = renderer.create(<Component />);

  expect(component.toJSON()).toMatchSnapshot();
});

test("renders DatePicker with custom input", () => {
  const Component = () => {
    return (
      <DatePicker
        customInput={<input className="custom-input" value="" />}
        selected={new Date()}
        onChange={() => {}}
      />
    );
  };

  const component = renderer.create(<Component />);

  expect(component.toJSON()).toMatchSnapshot();
});
