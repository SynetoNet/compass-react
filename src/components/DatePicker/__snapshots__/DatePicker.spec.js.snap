// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DatePicker with custom input 1`] = `
<div
  className="react-datepicker-wrapper"
>
  <div
    className="react-datepicker__input-container"
  >
    <input
      className=""
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      readOnly={false}
      value="23/05/2019"
    />
  </div>
</div>
`;

exports[`renders default DatePicker 1`] = `
<div
  className="react-datepicker-wrapper"
>
  <div
    className="react-datepicker__input-container"
  >
    <input
      className="form-control"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      readOnly={false}
      value="23/05/2019"
    />
  </div>
</div>
`;
